### PR TITLE
Reingest messages into Firehose

### PR DESCRIPTION
**Pull request by andylangridge**
* Reads events from the S3 > SQS Firehose failure queue
* Gets S3 key and bucket from event, and extracts the relevant audit events - added partial failure response to Lambda in case any of these calls fail - these can then just return to the queue and be retried
* Puts the audit events back onto Firehose in a batch - Handle any partial failures from Firehose Put Batch command - if there are any we just update the S3 object instead of deleting it (which triggers a new object created event).
* Delete S3 object when all events reingested

### CODE REVIEW

File: src/lambdas/auditMessageFirehoseReingest/getAuditEvents.test.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+  it('returns the audit events from the provided s3 objects', async () => {
+    when(getAuditEventsFromS3Object).mockResolvedValue(auditEvents)
+
+    const expectedResult = {
+      successfulResults: [
+        {
+          auditEvents,
+          bucket: 'mockBucket',
+          key: 'failures/mockKey1',
```
Comment (gdsandrewmccafferty) [inline src/lambdas/auditMessageFirehoseReingest/getAuditEvents.test.ts:37]: If you're setting up an assertion like this, I'd rather we didn't just copy and paste strings, and just set up some constants somewhere (probably just in the same file)

File: src/lambdas/auditMessageFirehoseReingest/getAuditEventsFromS3Object.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import { base64ToObject } from '../../utils/helpers/base64ToObject'
+import { readableToString } from '../../utils/helpers/readableToString'
+
+export const getAuditEventsFromS3Object = async (
+  bucketName: string,
+  key: string
+) => {
+  const contents = await getS3ObjectAsStream(bucketName, key)
+
+  return await getAuditEvents(contents)
```
Comment (gdsandrewmccafferty) [inline src/lambdas/auditMessageFirehoseReingest/getAuditEventsFromS3Object.ts:14]: I'm never quite sure what `return await` will do, but I think it indicates that you should probably just be returning the promise. Would be good to add return types for your functions to make it clearer what's actually going on here as well.

File: src/lambdas/auditMessageFirehoseReingest/deleteOrUpdateS3Objects.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import { objectToBase64 } from '../../utils/helpers/objectToBase64'
+
+export const deleteOrUpdateS3Objects = async (results: S3ObjectDetails[]) => {
+  return Promise.all(
+    results.map(async (result) => {
+      if (
+        result.auditEventsFailedReingest &&
+        result.auditEventsFailedReingest.length > 0
+      ) {
+        try {
```
Comment (gdsandrewmccafferty) [inline src/lambdas/auditMessageFirehoseReingest/deleteOrUpdateS3Objects.ts:16]: You could do with splitting some of this up into separate functions (within the same file, but just for readability). Having an `if` block followed by a `try/catch`, and then an `else` is a bit much and quite difficult to follow

File: src/lambdas/auditMessageFirehoseReingest/handler.test.ts [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    jest.clearAllMocks()
+  })
+
+  it('successfully receives an SQS event, and returns an SQS Batch response', async () => {
+    when(getAuditEvents).mockResolvedValue({
+      successfulResults: [
+        {
+          auditEvents: [],
+          bucket: 'mockBucket',
+          key: 'failures/your-object-key',
```
Comment (gdsandrewmccafferty) [inline src/lambdas/auditMessageFirehoseReingest/handler.test.ts:55]: A lot of repetition of the same string here, would be better to use a constant to save typos later on

File: src/lambdas/auditMessageFirehoseReingest/handler.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+  return {
+    batchItemFailures
+  }
+}
+
+const getS3ObjectDetails = (records: SQSRecord[]): S3ObjectDetails[] => {
+  const s3ObjectDetails = records
+    .filter((record) => {
+      const isS3TestEvent = tryParseJSON(record.body).Event === 's3:TestEvent'
```
Comment (gdsandrewmccafferty) [inline src/lambdas/auditMessageFirehoseReingest/handler.ts:61]: The logic for working out if this is a test event could go in a function to bring down the line count of this function a bit

File: src/lambdas/auditMessageFirehoseReingest/handler.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+      if (isS3TestEvent) {
+        logger.info('Event is of type s3:TestEvent, ignoring')
+      }
+
+      return !isS3TestEvent
+    })
+    .filter((record) => {
+      const s3EventData = tryParseJSON(record.body).Records[0].s3
```
Comment (gdsandrewmccafferty) [inline src/lambdas/auditMessageFirehoseReingest/handler.ts:70]: Again, the logic of working out if this a failure key could be its own function

File: src/lambdas/auditMessageFirehoseReingest/sendAuditEventsToFirehose.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+import { AuditEvent } from '../../types/auditEvent'
+import { S3ObjectDetails } from '../../types/s3ObjectDetails'
+import { auditEventsToFirehoseRecords } from '../../utils/helpers/firehose/auditEventsToFirehoseRecords'
+import { getEnv } from '../../utils/helpers/getEnv'
+
+export const sendAuditEventsToFirehose = async (
+  s3ObjectDetails: S3ObjectDetails[]
+): Promise<S3ObjectDetails[]> =>
+  Promise.all(
+    s3ObjectDetails.map(async (details) => {
```
Comment (gdsandrewmccafferty) [inline src/lambdas/auditMessageFirehoseReingest/sendAuditEventsToFirehose.ts:13]: Wherever possible, I'd prefer to be able to see exactly what's going on in a map with a nicely named function, think that's possible here

File: src/lambdas/auditMessageFirehoseReingest/getAuditEventsFromS3Object.ts (hunk 2 above)
Comment (andylangridge) [reply src/lambdas/auditMessageFirehoseReingest/getAuditEventsFromS3Object.ts]: Fixed. Don't need the `await` either!

File: src/lambdas/auditMessageFirehoseReingest/deleteOrUpdateS3Objects.ts (hunk 3 above)
Comment (andylangridge) [reply src/lambdas/auditMessageFirehoseReingest/deleteOrUpdateS3Objects.ts]: Have split up into functions now! Hope this is better?

File: src/lambdas/auditMessageFirehoseReingest/handler.ts (hunk 5 above)
Comment (andylangridge) [reply src/lambdas/auditMessageFirehoseReingest/handler.ts:61]: Ok filters moved. How is that?

File: src/lambdas/auditMessageFirehoseReingest/handler.ts (hunk 6 above)
Comment (andylangridge) [reply src/lambdas/auditMessageFirehoseReingest/handler.ts:70]: Filters moved.

File: src/lambdas/auditMessageFirehoseReingest/getAuditEvents.test.ts (hunk 1 above)
Comment (andylangridge) [reply src/lambdas/auditMessageFirehoseReingest/getAuditEvents.test.ts:37]: Done

File: src/lambdas/auditMessageFirehoseReingest/handler.test.ts (hunk 4 above)
Comment (andylangridge) [reply src/lambdas/auditMessageFirehoseReingest/handler.test.ts:55]: Done

File: src/lambdas/auditMessageFirehoseReingest/sendAuditEventsToFirehose.ts (hunk 7 above)
Comment (andylangridge) [reply src/lambdas/auditMessageFirehoseReingest/sendAuditEventsToFirehose.ts:13]: Done

**Review by gdsandrewmccafferty — APPROVED** [review]

**Pull request merged**